File: Lesson 5 - Prefix Sums/GenomicRangeQuery.cpp
#include <assert.h>
#include <vector>
#include <unordered_map>

std::vector<int> solution(std::string &S, std::vector<int> &P, std::vector<int> &Q)
{
    if (!S.size() || !P.size() || P.size() != Q.size())
    {
        return std::vector<int>();
    }

    std::unordered_map<char, int> nucleotidesImpact({ {'A', 0},{ 'C', 1 },{ 'G', 2 },{ 'T', 3 } });

    std::vector<int> DNA;
    DNA.reserve(S.size());
    for (const auto& nucleotide : S)
    {
        DNA.push_back(nucleotidesImpact[nucleotide]);
    }

    std::vector<std::vector<int>> factorsAmount(DNA.size(), std::vector<int>(4));
    std::vector<int> firstFactor(4);
    firstFactor[DNA[0]]++;
    factorsAmount[0] = firstFactor;

    for (size_t i = 1; i < DNA.size(); i++)
    {
        factorsAmount[i] = factorsAmount[i - 1];
        factorsAmount[i][DNA[i]]++;
    }

    std::vector<int> result(P.size());
    for (size_t i = 0; i < result.size(); i++)
    {
        if (P[i] > Q[i] || P[i] < 0 || Q[i] >= static_cast<int64_t>(S.size()))
        {
            continue;
        }

        const auto left = (P[i] - 1 >= 0) ? factorsAmount[P[i] - 1] : std::vector<int>(4);
        const auto right = factorsAmount[Q[i]];

        for (size_t j = 0; j < left.size(); j++)
        {
            if (right[j] > left[j])
            {
                result[i] = j + 1;
                break;
            }
        }
    }

    return result;
}

int main()
{
    assert(std::vector<int>({ 0, 0, 0, 1, 2, 4, 1 }) == solution(std::string("CAGCCTA"), std::vector<int>({ -1, 0, 5, 1, 2, 5, 0 }), std::vector<int>({ 6, 7, 4, 1, 4, 5, 6 })));
    assert(std::vector<int>({ 2, 1, 2 }) == solution(std::string(1000, 'C') + 'A' + std::string(1000, 'C'), std::vector<int>({ 0, 0, 1001 }), std::vector<int>({ 999, 2000, 2000 })));
}

File: Lesson 5 - Prefix Sums/GenomicRangeQuery.py
def solution(S, P, Q):
    if not len(S) or not len(P) or len(P) != len(Q):
        return []

    nucleotidesImpact = {key:value for key, value in zip(['A', 'C', 'G', 'T'], range(4))}
    DNA = []
    for nucleotide in S:
        DNA.append(nucleotidesImpact[nucleotide])

    firstFactor = [0] * 4
    firstFactor[DNA[0]] += 1
    factorsAmount = []
    factorsAmount.append(firstFactor)

    for i in range (1, len(DNA)):
        factorsAmount.append(factorsAmount[-1][:])
        factorsAmount[-1][DNA[i]] += 1

    result = [0] * len(P)
    for i in range(len(result)):
        if P[i] > Q[i] or P[i] < 0 or Q[i] >= len(S):
            continue

        left = factorsAmount[P[i] - 1] if (P[i] - 1 >= 0) else [0] * 4
        right = factorsAmount[Q[i]]

        for j in range(len(left)):
            if right[j] > left[j]:
                result[i] = j + 1
                break

    return result


assert [ 0, 0, 0, 1, 2, 4, 1 ] == solution("CAGCCTA", [ -1, 0, 5, 1, 2, 5, 0 ], [ 6, 7, 4, 1, 4, 5, 6 ])
assert [ 2, 1, 2 ] == solution('C' * 1000 + 'A' + 'C' * 1000, [ 0, 0, 1001 ], [ 999, 2000, 2000 ])
